### DISCUSSION TITLE
What if I wrote an frontend GUI client for Stash in windows?

### DISCUSSION THREAD

**Discussion by philpw99** [stashapp/stash#1854] [Ideas]
I wrote some small easy programs in AutoItScript. It's quite easy to write a GUI frontend in AutoIt for Stash. It will work like NZBget. Once run, it will launch the browser then hide in the Windows notification area (where other small icons are). You can click on it, a small menu will pop up, and you can choose the following actions:
* Go to the scenes, movies or performers section, or other specified lists.
* Install a scraper from a list of websites, or uninstall it.
* Install or uninstall plugins.
* Stop or restart stash.
* Install or uninstall css snippets, like "Enable NSFW filters".
* One emergency shotcut key to close all stash related browsers.
Those are the things I could think of for this little program. What do you guys think about? Please let me know. Thank you!

**Comment by kermieisinthehouse** [comment]
Something like this is already in the early stages of planning, and in native go so it can be compatible against all of our supported platforms. We need to do some more design work before we're ready to start cutting in features.

**Reply by philpw99** [reply]
Yes, I know this is all planned. Yet I am wondering if I can start an easy project for windows users. It will only take like 1 to 2 weeks to have the first version, then one month or two it will be fully ready.
I know it will not be easy to launch a GUI frontend for all the platforms.
I am just thinking, maybe a quick frontend for our windows users. Then once the official frontend is out, my little program can retire happily?

**Comment by WithoutPants** [comment]
This sounds more like a control app for stash rather than a front end (I would imagine a stash front end would have a similar interface to the web UI - I wouldn't want to maintain another separate UI for stash). 

This sounds like a useful utility. It's not something I'd personally use, and at this point not something I'd be keen to maintain personally, but if there's a desire for something like this then by all means give it a crack. We can always link to your project from the wiki/community scripts repo.

**Reply by philpw99** [reply]
Wonderful ! I do have a desire for something like that to manage my StashApp more easily. I will be maintaining the little program until you have your own nice GUI released.

**Comment by 7dJx1qP** [comment]
I think any step towards making the management of plugins, scrapers, and css easier and beginner friendly is a good idea.

Aside from development work required, is there any reason these features couldn't be put into the web UI? At least the installation and management of plugins, scrapers, and custom css seems like it would just need to check against the stash community repos and manage the user's plugins and scrapers folders.

**Reply by WithoutPants** [reply]
No, and I will be taking steps to improve these at some point - especially given that the management of scrapers (#623) is one of the top +1'd issues.

**Reply by philpw99** [reply]
Yup, I know those features will be built-in feature of Stash soon, but given you guys already have a long long to-do list, I think a small program might help. There are always some guys asking, "How do I do this do that..." You can just tell them to get my GUI and let them be quiet.
I am sure the real Stash GUI will be much better than mine, and it will be cross platform, but I guess that will be months away. This little thing will take a couple of weeks only. Just an easy fix.

**Comment by philpw99** [comment]
Just one day, some basic functions are done. Next obstacle is WebDriver. Not that easy. Wish me luck! 😅

**Reply by philpw99** [reply]
WebDriver is done. It works like magic ! 😄  I thought the first version will take at least two weeks, but now it seems it can be much quicker !

**Comment by philpw99** [comment]
This little program now is getting closer and closer to the first official release ! 😅 
Some screen shots for ya all.

This is the welcome screen + initial settings.

![WelcomeScreen](https://user-images.githubusercontent.com/22040708/137665773-737f9006-de02-4b93-8f25-127f37d3cdae.png)

Now once it runs, it will run stash-win.exe and webdriver.exe in the tray area, and a browser will be opened.
Below is the tray menu it offers.

![Helper2](https://user-images.githubusercontent.com/22040708/138910548-f7131c8e-794e-4aa9-becc-6ee238670569.png)

As you can see, all the sub-menus can be customize to launch whatever you like to see. I think 11x8 slots should be enough.
To customize it, choose the "Custom..." item, and you will see:

![CustomScenes](https://user-images.githubusercontent.com/22040708/137666178-463de459-558e-4299-8820-5a62ee99fc69.png)

You can also change the simple settings. Notice that there is a "Boss Coming Key" available. 😸 

![Helper3](https://user-images.githubusercontent.com/22040708/138910775-aa8c856f-f6a7-47c1-9de1-d25aff78c472.png)

Last and the best, but not yet finished, is the holy scraper list:

![helper6](https://user-images.githubusercontent.com/22040708/138910898-a92a47a9-756d-4415-bb0b-5c83a0508460.png)

I hope you like what you see here. Have spent the whole weekend just to create all of the above. I guess the first GitHub release will be this week.

**Reply by philpw99** [reply]
First release ! 😄 

https://github.com/philpw99/Stash_Helper/releases/tag/v1.0

**Comment by philpw99** [comment]
* [Stash_Helper 1.2 ](https://github.com/philpw99/Stash_Helper/releases)is out ! The big feature is "Using external media player to play the current scene". Now I can use PotPlayer to play the scene that I choose.
* The scraper manager is done. It's primitive but functional. It will allow downloading "SCRAPER-LIST.md" only once in a day. This way the loading of the list will be faster next time, and less stress on the GitHub server.
* Also I made a long [readme.md](https://github.com/philpw99/Stash_Helper/blob/main/README.md) for my helper, which may help Stash beginners to start their first installation. Please take a look and tell me what you think. Thank you !

**Reply by philpw99** [reply]
Hm... Now the version 1.6 is out. This program is getting better and stabler over the time. I am glad the progress I made so far:
* v1.3 improve the Scraper Manager.
* v1.4 can play movies and all the scenes under that movie.
* v1.5 added the update scrapers feature. Also the one-click "Scan New Files" in the tray menu.
* v1.6 added the feature of creating a movie from a scene and copy all the info over. It can also batch create movies for scenes that yet to have a movie link.
And of course, many small bugs are eliminated during the process.

To-Do list for this weekend:

- [x] Create a movie from the current scene.
- [x] Make play lists for external players.

**Reply by philpw99** [reply]
The v1.7 should be the one that counts as a final release. It adds a new "Play List" feature, which allows you to add scenes or movies to an internal play list, then you can save or load the list. You can also send the list to your favorite media player and let them play the files. The play list is in standard m3u format, so most media players should have no problem handling it.
Now I had achieved all the goals that I set out to do, don't know what more I can do to improve it. If you have an idea, please reply here and let me know. Thank you !

**Comment by philpw99** [comment]
v1.8 is out, based on the suggestions BViking78 made. Some keyboard shortcuts were added for the playlist feature:
    Ctrl-Alt-A(dd) for adding a scene or a movie to the list.
    Ctrl-Alt-C(lear) to clear the current play list.
    Ctrl-Alt-M(anage) to manage the list.
    Ctrl-Alt-P(lay) to play the list in external media player.

Also a "Bookmark" menu item was added, shortcut is Ctrl-Alt-B, for quickly adding the currently browsing Stash content to the menu.

For example, if you are looking at scenes using a filter "Is Missing: Performers", you hit Ctrl-Alt-B, Stash_helper will ask you to give it some title, you type in "Missing performers" and hit enter. The URL for this list will be added to the Stash_helper's "Scenes" list. Next time you just click on that "Missing performers" menu item, it will open to the same scenes list.

In a way, Bookmark button is like the bookmarks in web browsers. It's just more organized and convenient.

**Comment by philpw99** [comment]
v2.0 is finally, finally out. There were so many things needed to take care of, just to implement one feature !
In this version, you can filter scenes/images/movies/galleries to the ones you like to play, then "Ctrl-Alt-A" to add them to the list.
Previously images and galleries were not supported, and you had to add the scenes/movies one by one.
Now adding to the list is much easier. You can also define the order to add the list as well. Basically it will follow your browser, if the browser says "108 scenes", then you will add all those 108 scenes to the list. I don't follow the pagination on purpose.
Because of this new feature, sending a list of images to the media player is like piece of a cake. However, the media player always play the images with their own timing. So even the playlist says 5 seconds, the player might still play them at 10 seconds each. 
In addition, VLC is horrible at handling the long names. PotPlayer is much better.

**Comment by philpw99** [comment]
v2.1 add a feature called "CSS Magic". It's surprisingly easy to implement. I should have done it long time ago!
Instantly blur/unblur all the images with the NSFW css filter ? Haha, you bet. Instant CSS Magic !

**Comment by philpw99** [comment]
v2.2 is so good that I want to add it here!
Now you can view a gallery, well, just like a gallery !
![screen2](https://user-images.githubusercontent.com/22040708/147072712-e247518f-0544-4f97-8274-c5aa107c6bdd.png)
With this command:
![screen1](https://user-images.githubusercontent.com/22040708/147072222-e81051f6-f8ab-43dd-ae90-7f2f95ee30a8.png)
You can view a gallery like above. It's really fantastic for watching images and galleries in Stash. Love this addition !

**Reply by philpw99** [reply]
Sorry! Most of the v2.2.1 queries are broken. I only knew it when I tried to add movies to the playlist.
Don't use v2.2.1, I will delete that release. v.2.2.2 is the right one.
